pkg/ext/apiserver.go

```diff
@@ -0,0 +1,276 @@
+	"k8s.io/apiserver/pkg/endpoints/openapi"
+	"k8s.io/apiserver/pkg/endpoints/request"
+	"k8s.io/apiserver/pkg/registry/rest"
+	genericapiserver "k8s.io/apiserver/pkg/server"
+	genericoptions "k8s.io/apiserver/pkg/server/options"
+	utilversion "k8s.io/apiserver/pkg/util/version"
+	"k8s.io/client-go/kubernetes"
+	openapicommon "k8s.io/kube-openapi/pkg/common"
+	"k8s.io/kube-openapi/pkg/validation/spec"
+)
+
+var (
+	// Reference:
+	// Scheme scheme for unversioned types - such as APIResourceList, and Status
+	scheme = runtime.NewScheme()
+	Scheme = scheme
+	// Codecs for unversioned types - such as APIResourceList, and Status
+	Codecs = serializer.NewCodecFactory(scheme)
+)
+
+func init() {
+	schemes.AddToScheme(scheme)
+	metav1.AddToGroupVersion(scheme, schema.GroupVersion{Version: "v1"})
+	metainternalversion.RegisterConversions(scheme)
+
+	unversioned := schema.GroupVersion{Group: "", Version: "v1"}
+	scheme.AddUnversionedTypes(unversioned,
+		&metav1.Status{},
+		&metav1.APIVersions{},
+		&metav1.APIGroupList{},
+		&metav1.APIGroup{},
+		&metav1.APIResourceList{},
+	)
+}
+
+type ExtensionAPIServerOptions struct {
+	// GetOpenAPIDefinitions is collection of all definitions. Required.
+	GetOpenAPIDefinitions             openapicommon.GetOpenAPIDefinitions
+	OpenAPIDefinitionNameReplacements map[string]string
+
+	Authentication AuthenticationOptions
+
+	Authorization authorizer.Authorizer
+
+	Client kubernetes.Interface
+
+	BindPort int
+}
+
+type AuthenticationOptions struct {
+	// When turned off, reject requests, unless custom authentication passes
+	EnableBuiltIn       bool
+	CustomAuthenticator authenticator.RequestFunc
+}
+
+type ExtensionAPIServer struct {
+	codecs serializer.CodecFactory
+	scheme *runtime.Scheme
+
+	genericAPIServer *genericapiserver.GenericAPIServer
+	apiGroups        map[string]genericapiserver.APIGroupInfo
+
+	handlerMu sync.RWMutex
+	handler   http.Handler
+}
+
+type emptyAddresses struct{}
+
+func (e emptyAddresses) ServerAddressByClientCIDRs(clientIP net.IP) []metav1.ServerAddressByClientCIDR {
+	return nil
+}
+
+func NewExtensionAPIServer(scheme *runtime.Scheme, codecs serializer.CodecFactory, opts ExtensionAPIServerOptions) (*ExtensionAPIServer, error) {
+	recommendedOpts := genericoptions.NewRecommendedOptions("", codecs.LegacyCodec())
+
+	if opts.Authorization == nil {
+		return nil, fmt.Errorf("no authorization provided")
+	}
+
+	resolver := &request.RequestInfoFactory{APIPrefixes: sets.NewString("apis", "api"), GrouplessAPIPrefixes: sets.NewString("api")}
+	config := genericapiserver.NewRecommendedConfig(codecs)
+	config.RequestInfoResolver = resolver
+	config.Authorization = genericapiserver.AuthorizationInfo{
+		Authorizer: opts.Authorization,
+	}
+
+	// XXX: kubectl doesn't show this, and it's listed as optional so it
+	//      should be safe to remove. I haven't found a lot of resource on
+	//      what this does.
+	config.DiscoveryAddresses = emptyAddresses{}
```

Comment:
Did a bit more digging into this.

**The what**

The `DiscoveryAddresses` makes a list of endpoints to reach the extension API server based on the client CIDR. For example, you could have a public endpoint (eg: `mycluster.example.com:6443`)  as the default, but have a clusterIP endpoint (eg: `10.42.0.2:6443`) for clients within the cluster. Here's what it would look like from the discovery endpoint:

```json
{
  "kind": "APIGroupList",
  "apiVersion": "v1",
  "groups": [
    {
      "name": "ext.cattle.io",
      "versions": [
        {
          "groupVersion": "ext.cattle.io/v1",
          "version": "v1"
        }
      ],
      "preferredVersion": {
        "groupVersion": "ext.cattle.io/v1",
        "version": "v1"
      },
      "serverAddressByClientCIDRs": [
        {
          "clientCIDR": "0.0.0.0/0",
          "serverAddress": "mycluster.example.com:6443"
        },
        {
          "clientCIDR": "10.42.0.0/24",
          "serverAddress": "10.43.0.1:6443"
        }
      ]
    },
    ...
}
```

In practice, by going through the main kube-apiserver with k3s, I don't see those `serverAddressByClientCIDRs` being listed in `/apis` but I do see them in `/api`.

**The why**

From [the comments](https://github.com/kubernetes/apimachinery/blob/37988e577e160bf7749437141fee1d6b8ca15dd6/pkg/apis/meta/v1/types.go#L1109-L1118) and that [very small Slack conversation](https://kubernetes.slack.com/archives/C0EG7JC6T/p1550616735204100), that feature is there as an optimization so that clients can contact directly an extension API server without being proxied by k8s.

As an example from the previous link, one would be able to go directly to the metrics server. 

(**Note**: That slack conversation is old ~2019, it's unclear if the feature has been completed and if its used at all. I started a [new Slack thread](https://kubernetes.slack.com/archives/C0EG7JC6T/p1728325393372749) to get some more information.)

---

For Rancher, I think it's fine if the only client goes through the main kube-apiserver always. As such, we can either:
- figure out what our address is (I'm thinking `api-extension.cattle-system.svc:6443` would do, that would be a setting in steve) 
- keep it empty as is done here

I tried both and in both cases the server address isn't listed when running `kubectl get --raw /apis`.

Reply:
Overall, given that we will allow two methods of using this endpoint (the rancher token endpoint and the api-extension compatible method), I actually think that this would be a good thing for us to support. That being said:
- The alternative IPs and Hostnames should be supplied by the consuming code
- This is an optimization, not a required feature.

For those reasons I think that we can hold off on this for the moment.